dashboard/app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -5,6 +5,46 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
 
   # GET /users/auth/:provider/callback
   def all
+    auth_params = request.env['omniauth.params']
+    if auth_params['state'] == 'connect_provider' && current_user
+      connect_provider
+    else
+      login
+    end
+  end
+
+  # Call /users/auth/google_oauth2?state=connect_provider and the callback will trigger this code path
+  def connect_provider
+    # TODO: some of this won't work right for non-Google providers, because info comes in differently
+    auth_hash = request.env['omniauth.auth']
+    provider = auth_hash.provider.to_s
+    if AuthenticationOption::OAUTH_CREDENTIAL_TYPES.include? provider
+      new_data = nil
+      if auth_hash.credentials && (auth_hash.credentials.token || auth_hash.credentials.expires_at || auth_hash.credentials.refresh_token)
```

Comment:
I did this in the migrate method, but it might be cleaner to give AuthenticationOption a before_save or before_validate hook that turns an empty data hash into nil.